[#172099629] Allow creation from snapshot before date

tlwr opened the pull request:
[Story](https://www.pivotaltracker.com/story/show/172099629)

What
----

Currently when creating a new database from the snapshot of an existing one, only the last snapshot can be chosen

This PR adds a new flag, which can be used when creating from an existing snapshot:

```
# I ran this my dev env
cf create-service \
  postgres \
  tiny-unencrypted-11 \
  tlwr-test-snapshot \
  -c '{
    "restore_from_latest_snapshot_before": "2020-04-01 13:00:00", 
    "restore_from_latest_snapshot_of": "35a738c0-e53f-45f0-bf9f-f5a8f197b86d"
  }'
```

`restore_from_latest_snapshot_before` is optional, and only works with `restore_from_latest_snapshot_of`.

IMO this is a really nice UX (originally suggested by Andy Hunt)

How to review
---

- [ ] Code review
- [ ] Check `admin/noodle` where there are two databases created from snapshot one with `restore_from_latest_snapshot_before` and one without
- [ ] Check the RDS console in London for the two databases, and see that the tags are correctly set

Commit:
Add restore_from_latest_snapshot_before parameter

When provisioning with this parameter we will be able to select the
latest backup before a certain date

Signed-off-by: Toby Lorne <toby.lornewelch-richards@digital.cabinet-office.gov.uk>

Commit:
Test restore_from_latest_snapshot_before param

This doesnt implement restore_from_latest_snapshot_before but adds new
tests which test the proposed functionality

The diff is mainly wrapping the existing tests in a new Context
statement, to take advantage of BeforeEach and JustBeforeEach

Signed-off-by: Toby Lorne <toby.lornewelch-richards@digital.cabinet-office.gov.uk>

Commit:
Implement restore_from_latest_snapshot_before

When creating a new instance, we should be able to choose the latest
date/time which is acceptable to restore from

Signed-off-by: Toby Lorne <toby.lornewelch-richards@digital.cabinet-office.gov.uk>

AP-Hunt commented on rdsbroker/broker.go:
```diff
@@ -209,9 +215,35 @@ func (b *RDSBroker) Provision(
 			return brokerapi.ProvisionedServiceSpec{}, fmt.Errorf("No snapshots found for guid '%s'", *provisionParameters.RestoreFromLatestSnapshotOf)
 		}
+
 		snapshot := snapshots[0]
```
Could we add a log line show which snapshot is being used, since now it isn't necessarily the one most recent to the time of the other log lines?

tlwr replied:
This is captured in a tag on the subsequently created RDS instance: `Restored From Snapshot`

AP-Hunt replied:
What about in the event that database creation fails? I think if I was debugging that, I'd like to be able see in the logs which snapshot it picked

tlwr replied:
Good shout, added in e4f482944eb2e1bd320b8a5577c242a5389242ea

Commit:
Add snapshot pruning logging diagnostics

Suggestion from Andy:

> What about in the event that database creation fails?
> I think if I was debugging that, I'd like to be able see in the logs which snapshot it picked

Signed-off-by: Toby Lorne <toby.lornewelch-richards@digital.cabinet-office.gov.uk>
